ckanext/external_storage/blueprints.py

```diff
@@ -1,63 +1,39 @@
 """ckanext-external-storage Flask blueprints
 """
-from ckan import model
-from ckan.lib import uploader
 from ckan.plugins import toolkit
-from flask import Blueprint, send_file
+from flask import Blueprint
+
+from .download_handler import call_download_handlers, call_pre_download_handlers, get_context
 
 blueprint = Blueprint(
     'extstorage',
     __name__,
 )
 
 
-def download(resource_id, filename=None, **__):
-    """Redirect to an LFS provided download URL or fall back to default
-    download method if resource is not managed by External Storage
+def download(id, resource_id, filename=None):
+    """Download resource blueprint
+
+    This calls all registered download handlers in order, until
+    a response is returned to the user
     """
-    context = {
-        'model': model,
-        'user': toolkit.c.user,
-        'auth_user_obj': toolkit.c.userobj,
-    }
+    context = get_context()
+
     try:
         resource = toolkit.get_action('resource_show')(context, {'id': resource_id})
+        if id != resource['package_id']:
+            return toolkit.abort(404, toolkit._('Resource not found belonging to package'))
+
+        package = toolkit.get_action('package_show')(context, {'id': id})
+
+        resource = call_pre_download_handlers(resource, package)
+        return call_download_handlers(resource, package, filename)
+
     except toolkit.ObjectNotFound:
         return toolkit.abort(404, toolkit._('Resource not found'))
     except toolkit.NotAuthorized:
-        return toolkit.abort(401, toolkit._('Unauthorized to read resource {0}'.format(resource_id)))
-
-    if resource.get('url_type') == 'upload' and resource.get('lfs_prefix'):
-        # We are in LFS land!
-        return redirect_to_external_storage(context, resource, filename)
-    return fallback_download_method(resource)
+        return toolkit.abort(401, toolkit._('Not authorized to read resource {0}'.format(resource_id)))
```

Comment:
I think the core behavior is to return 404 on not authorized to not disclose if the package/resource exists or not. I don't care that much about it but flagging in case you want to align with it

Reply:
Ok; I think I took this from ckanext-cloudstorage but I will double check and fix in accordance.

Reply:
This behavior is copied from Cloudstorage, so I think I should keep it for now. Maybe The `package_show` API, as far as I can tell, also raises 401 for something that exists but is not accessible